src/CMakeLists.txt

```diff
@@ -629,30 +629,44 @@ install(TARGETS ceph-mon DESTINATION bin)
 # OSD/ObjectStore
 # make rocksdb statically
 if(NOT ALLOCATOR STREQUAL "jemalloc")
-  set(disable_jemalloc "DISABLE_JEMALLOC=1")
+  set(ROCKSDB_JEMALLOC "DISABLE_JEMALLOC=1")
 endif()
 
-set(ROCKSDB_EXTRA_CXXFLAG -fPIC -Wno-unused-variable)
+set(ROCKSDB_EXTRA_CXXFLAG "-fPIC -Wno-unused-variable")
 if(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
-  set(ROCKSDB_EXTRA_CXXFLAG ${ROCKSDB_EXTRA_CXXFLAG} -DNDEBUG)
+  set(ROCKSDB_EXTRA_CXXFLAG "${ROCKSDB_EXTRA_CXXFLAG} -DNDEBUG")
 endif(NOT CMAKE_BUILD_TYPE STREQUAL Debug)
 
 set(ROCKSDB_CXX "${CMAKE_CXX_COMPILER}")
 if (WITH_CCACHE AND CCACHE_FOUND)
-  set(ROCKSDB_CXX "ccache" "${CMAKE_CXX_COMPILER}")
+  set(ROCKSDB_CXX "ccache ${CMAKE_CXX_COMPILER}")
 endif(WITH_CCACHE AND CCACHE_FOUND)
 
-add_custom_target(build_rocksdb
-    COMMAND
-    PORTABLE=1 ${disable_jemalloc} $(MAKE) static_lib CXX="${ROCKSDB_CXX}" EXTRA_CXXFLAGS="${ROCKSDB_EXTRA_CXXFLAG}"
-    WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}/src/rocksdb
-    COMMENT "rocksdb building")
+set(ROCKSDB_AR "${CMAKE_AR}")
+
+# we use an external project and copy the sources to bin directory to ensure
+# that object files are built outside of the source tree.
+include(ExternalProject)
+ExternalProject_Add(rocksdb_ext
+    URL "${CMAKE_SOURCE_DIR}/src/rocksdb"
+    PREFIX "external"
+    PATCH_COMMAND ""
+    UPDATE_COMMAND ""
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND PORTABLE=1 ${ROCKSDB_JEMALLOC} $(MAKE) CXX=${ROCKSDB_CXX} AR=${ROCKSDB_AR} EXTRA_CXXFLAGS=${ROCKSDB_EXTRA_CXXFLAG} static_lib
+    BUILD_IN_SOURCE 1
+    INSTALL_COMMAND "")
+
+# force rocksdb make to be called on each time
+ExternalProject_Add_Step(rocksdb_ext forcebuild DEPENDEES configure DEPENDERS build ALWAYS 1)
+
+ExternalProject_Get_Property(rocksdb_ext source_dir)
```

Comment:
unfortunately there is no option to change the property name at least not in cmake 2.8.11 which is still our minimum version :-(